routers/api/v1/repo/branch.go

```diff
@@ -182,6 +182,87 @@ func DeleteBranch(ctx *context.APIContext) {
 	ctx.Status(http.StatusNoContent)
 }
 
+// CreateRepoBranch creates a branch for a user's repository
+func CreateRepoBranch(ctx *context.APIContext, opt api.CreateBranchRepoOption) {
+	err := repo_module.CreateNewBranch(ctx.User, ctx.Repo.Repository, opt.OldBranchName, opt.BranchName)
```

Comment:
Unfortunately CreateNewBranch doesn't do all the checks we need here.

If you're renaming the default branch for the repository you need to change that on the repository when you change the branch

If the branch is protected we probably shouldn't allow its name to be changed.

(I'm not sure if we're handling this appropriately elsewhere if not we should add the checking to repo_module.CreateNewBranch and probably consider moving the checks there if so.)

Reply:
I might be missing something but does CreateNewBranch rename a branch ?

It does create a new branch from an "old" one (the "OldBranchName" name is a bit misleading) but does not rename it, does it ?

If I understood the issue correctly, we want to be able to create a new branch from another one. Do we need to change the Repository's BranchName ?

Reply:
Oh sorry I just got completely confused by what you were trying to do!!